rfcs/0138-rfc-repos.md

```diff
@@ -0,0 +1,65 @@
+---
+feature: rfc-repos
+start-date: 2022-01-10
+author: Silvan Mosberger
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Develop and discuss RFCs in repositories instead of a pull request.
+
+# Motivation
+[motivation]: #motivation
+
+RFC discussions are currently held in single pull requests on the [rfcs](https://github.com/NixOS/rfcs/pulls) repository.
+Due to GitHub's limitations on pull requests, this quickly becomes unmanageable.
+The most prominent example of this is [the Flakes RFC](https://github.com/NixOS/rfcs/pull/49) with over 500 comments.
+
+# Detailed design
+[design]: #detailed-design
+
+The RFC process is changed to use repositories to develop and review RFCs instead.
+To create a new RFC, an issue on the [rfcs](https://github.com/NixOS/rfcs) repository is opened.
+It should contain the following contents describing the process:
+
+> This RFC is developed as the README.md of this repository: https://github.com/$owner/$repo
+>
+> To review the RFC
+> - Open https://github.com/$owner/$repo/blob/master/README.md?plain=1
+> - Select the lines you wish to comment on
+> - Click the "..." menu on the left
+> - Select "Reference in new issue" to open a new issue commenting on those lines
+>
+> To suggest changes, open a PR against the repository.
+>
+> All issues and pull requests must be resolved before the FCP can be initiated and completed.
+>
+> This issue may only be used for RFC meta-discussions, such as shepherd nominations, FCP periods, meeting schedules, etc.
+
+The FCP must be initiated on a specific commit of the repository.
+When the FCP passes, the repositories contents are committed to the [rfcs](https://github.com/NixOS/rfcs) repository by the [RFC steering committee](https://github.com/NixOS/rfcs#rfc-steering-committee) via a pull request that closes the original issue.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+
+# Alternatives
```

Comment:
I suggest putting more focus on Pre-RFCs and on discussion in the forum. Rust does this and it works really well IMO. A lot of the discussion noise in GitHub threads currently comes to RFCs being posted as drafts containing typos (I've also occasionally witnessed this on Matrix RFCs too), or more generally from the first impact of an idea in the author's head with the reality of a community's opinions. A lot of that can be absorbed by iterating outside of the formal RFC process. I think then the discussion workload on GitHub would remain manageable.

Reply:
Yeah that's a good suggestion. I think these two ideas go hand-in-hand: You can create a repository for development of the RFC beforehand, and if you want it to become more official, be looked at by more people, etc. open an issue in the rfcs repository linking to the repository where it's being developed.

Reply:
Maybe then a ful alternative could look like that:

* in the RFC process: add a field «acknowledgements» to be used mainly for reviewers of early drafts who were only giving feedback and local suggestions and not co-writing the text; clarify these people are eligible to be shepherds
* in the RFC process: establish that representing different viewpoints based on different use patterns in a shepherd team is desired, but only if different viewpoints are actually expressed during the early formal review
* informally: encourage early direction-discussion on Discourse
* informally: encourage splitting the concerns for making progress step by step
* informally: if the minimal useful version is still rather large, encourage a round of drafting in a separate repository before submission, adding the active participants into early reviewers (well, possibly some will also become authors)

I do think that splitting small enough not to need a separate repo is good when possible, and I think that submitting for the final review via the old process after drafting in a repo might be still reasonable otherwise.